File: DiscordBot/Modules/LeaderBoard.cs
using System.Net.Http.Json;
using System.Reflection.Metadata;
using System.Runtime.InteropServices.JavaScript;
using System.Text;
using System.Text.Json;
using Discord;
using Discord.Commands;
using Discord.Utils;
using Discord.Interactions;
using Discord.Interactions.Builders;
using DiscordBot.Services;
using ESportLeaderBoard.Model;
using Microsoft.AspNetCore.Hosting;
using RequireUserPermissionAttribute = Discord.Commands.RequireUserPermissionAttribute;
using Game = ESportLeaderBoard.Model.Game;

namespace DiscordBot.Modules;

public class LeaderBoard : InteractionModuleBase<SocketInteractionContext>
{
    private HttpClient _client;
    public InteractionService Commands { get; set; }
    private CommandHandler _handler;

    public LeaderBoard (CommandHandler handler)
    {
        _handler = handler;
        _client = new HttpClient();
    }
    [SlashCommand("initplayers","Init User List on LeaderBoard API")]
    [RequireUserPermission(GuildPermission.Administrator)]
    public async Task InitPlayers()
    {
        Console.WriteLine("InitPlayers");
        var members = Context.Guild.Users.Where(u => !u.IsBot).ToList();
        var responsePlayers =
            members.Select(u => new {name = u.DisplayName, profilePicture = u.GetDisplayAvatarUrl()}).ToList();
        var json = JsonSerializer.Serialize(responsePlayers);
        var output= await _client.PutAsync(@"http://localhost:5001/Player", new StringContent(json, Encoding.UTF8, "application/json"));
        await Context.Interaction.RespondAsync($"```css\n{output.StatusCode}-{(await output.Content.ReadAsStringAsync())}```");
        
    }

    [SlashCommand("showleaderboard", "Show Leaderboard")]
    public async Task ShowLeaderboard(Game game)
    {
        var board = await (await _client.GetAsync($"http://localhost:5001/LeaderBoard/{game}")).Content.ReadFromJsonAsync<LeaderBoardResponse>();
        if(board == null)
        {
            await Context.Interaction.RespondAsync($"```css\n LeaderBoard {game.ToString()} not found!```");
            return;
        }
        if(board.Players is null)
        {
            await Context.Interaction.RespondAsync($"```css\n LeaderBoard {game.ToString()} is empty!```");
            return;
        }
        var embed = CreateFields(board);
        /*await Context.Channel.SendMessageAsync($"```css\n Loading {game.ToString()} LeaderBoard...```")
            .ContinueWith(async t =>
            {
                await Task.Delay(2000);
                await Context.Channel.ModifyMessageAsync(t.Result.Id, m => m.Embeds = new[] {embed.Build()});
            });*/
        await Context.Interaction.RespondAsync(embed: embed);
    }
    public Embed CreateFields(LeaderBoardResponse board)
    {
        var fields = board.Players
            .Select((p,i) => new EmbedFieldBuilder
            {
                Name =  $"{i+1:###}.  |   {p.Player.Name} {new string(' ',30-CountChars(p.Player.Name))} **{p.Score}**",
                Value = new string('-',30),
                IsInline = false
            }).ToList();
        
        var embed = new EmbedBuilder()
            .WithTitle($"*{board.Game.ToString()}*")
            .WithColor(Color.Blue)
            .WithFields(fields);
        return embed.Build();
    }
    
    private int CountChars(string s)
        => s.Count(c => char.ToLower(c) >= ' ' && char.ToLower(c) <= '~');
    
    /*
    public static List<ApplicationCommandOptionChoiceProperties> CreateChoices()
    {
        var httpClient = new HttpClient();
        var response = httpClient.GetAsync(@"http://localhost:5001/LeaderBoard/Types").Result;
        if(!response.IsSuccessStatusCode)
        {
            throw new Exception($"Error while getting LeaderBoard Types: {response.StatusCode}-{response.ReasonPhrase}");
        }
        return response.Content.ReadFromJsonAsync<string[]>().Result?.Select(s =>
               {
                     return new ApplicationCommandOptionChoiceProperties
                     {
                         Name = s,
                            Value = s
                            
                     };
               }).ToList() 
               ?? new List<ChoiceAttribute>();
    }*/
}

File: DiscordBotpy/cogs/LeaderBoard.py
import subprocess

import discord_slash.model
from discord.ext import commands
import json
import discord
import os
from discord_slash.utils.manage_components import create_button, create_actionrow
from discord_slash.model import ButtonStyle
import requests
from discord_slash import SlashCommand,SlashContext, cog_ext
from discord_slash.model import SlashCommandOptionType
from discord_slash.utils.manage_commands import create_choice,create_option

def GetGames() -> list:
    out = requests.get(f"http://localhost:5001/LeaderBoard/Types", headers={'Content-Type': 'application/json'},)
    if not out.ok:
        raise Exception(f"Error: {out.status_code}-{out.reason}")
    return list(map(lambda x: create_choice(name=x,value=x),out.json()))


class LeaderBoard(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.leaderboards = []

    @cog_ext.cog_slash(
        name='initplayers',
        description='Init User List on LeaderBoard API')

    @commands.has_permissions(administrator = True)
    async def InitPlayers(self, ctx : SlashContext, ip : str):
        members = list(filter(lambda m: not m.bot, ctx.guild.members))
        requestPlayers = list(map(lambda x: {"name":x.display_name,"profilePicture":str(x.avatar_url)},members))
        out = requests.put(f"http://localhost:5001/Player", json=requestPlayers,  headers={'Content-Type': 'application/json'}, )
        await ctx.send(f"```css\n{out.status_code}-{out.reason}```")

    @cog_ext.cog_slash(
        name='showleaderboard',
        description='Show LeaderBoard',
        options=[
            create_option(
                name="board",
                description="Witch LeaderBoard do you want to see?",
                option_type=SlashCommandOptionType.STRING,
                required=False,
                choices=GetGames()
            )
        ])
    @commands.has_permissions(administrator = True)
    async def ShowLeaderBoard(self, ctx : SlashContext, board : str):
        self.leaderboards.append(ctx.message)
        await ctx.send("Loading LeaderBoard...")


def setup(bot):
    bot.add_cog(LeaderBoard(bot))